File: go/142.linked-list-cycle-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
    return twoPointers(head)
}

func twoPointers(head *ListNode) *ListNode {
    i := findIntersection(head)
    if i == nil {
        return nil
    }
    j := head
    for i != j {
        i = i.Next
        j = j.Next
    }
    return i
}

func findIntersection(head *ListNode) *ListNode {
    i := head
    j := head
    for i != nil && j != nil {
        i = i.Next
        j = j.Next
        if j != nil {
            j = j.Next
        }
        if i == j {
            return i
        }
    }
    return nil
}

File: csharp/142.linked-list-cycle-ii.cs
/*
 * @lc app=leetcode id=142 lang=csharp
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution 
{
    public ListNode DetectCycle(ListNode head) 
    {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null)
        {
            fast = fast.next.next;
            slow = slow.next;

            // https://www.geeksforgeeks.org/floyds-cycle-finding-algorithm/
            // cycle detected
            if (fast == slow)
            {
                // reset any pointer
                slow = head;
                while (fast != slow)
                {
                    slow = slow.next;
                    fast = fast.next;
                }
                return slow;
            }
        }

        return null;
    }
}
// @lc code=end

